compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/generators/DelegatedMemberGenerator.kt

```diff
@@ -0,0 +1,260 @@
+import org.jetbrains.kotlin.fir.declarations.builder.buildProperty
+import org.jetbrains.kotlin.fir.declarations.builder.buildSimpleFunction
+import org.jetbrains.kotlin.fir.declarations.builder.buildValueParameterCopy
+import org.jetbrains.kotlin.fir.declarations.impl.FirDeclarationStatusImpl
+import org.jetbrains.kotlin.fir.symbols.CallableId
+import org.jetbrains.kotlin.fir.symbols.impl.FirNamedFunctionSymbol
+import org.jetbrains.kotlin.fir.symbols.impl.FirPropertySymbol
+import org.jetbrains.kotlin.fir.symbols.impl.FirVariableSymbol
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.declarations.impl.IrFunctionImpl
+import org.jetbrains.kotlin.ir.declarations.impl.IrPropertyImpl
+import org.jetbrains.kotlin.ir.declarations.impl.IrValueParameterImpl
+import org.jetbrains.kotlin.ir.descriptors.WrappedPropertyDescriptor
+import org.jetbrains.kotlin.ir.descriptors.WrappedSimpleFunctionDescriptor
+import org.jetbrains.kotlin.ir.descriptors.WrappedValueParameterDescriptor
+import org.jetbrains.kotlin.ir.expressions.impl.*
+import org.jetbrains.kotlin.ir.types.classOrNull
+import org.jetbrains.kotlin.ir.types.impl.IrSimpleTypeImpl
+import org.jetbrains.kotlin.ir.types.isNothing
+import org.jetbrains.kotlin.ir.types.isUnit
+import org.jetbrains.kotlin.ir.util.defaultType
+import org.jetbrains.kotlin.ir.util.hasAnnotation
+import org.jetbrains.kotlin.ir.util.isFakeOverride
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.resolve.DescriptorUtils
+
+/**
+ * A generator for delegated members from implementation by delegation.
+ *
+ * It assumes a synthetic field with the super-interface type has been created for the delegate expression. It looks for delegatable
+ * methods and properties in the super-interface, and creates corresponding members in the subclass.
+ * TODO: generic super interface types and generic delegated members.
+ */
+internal class DelegatedMemberGenerator(
+    private val components: Fir2IrComponents
+) : Fir2IrComponents by components {
+
+    // Generate delegated members for [subClass]. The synthetic field [irField] has the super interface type.
+    fun generate(irField: IrField, subClass: IrClass) {
+        (irField.type as IrSimpleTypeImpl)?.classOrNull?.owner?.declarations?.filter {
+            it.isDelegatable() && !subClass.declarations.any { decl -> isOverriding(irBuiltIns, decl, it) }
+        }?.map {
+            if (it is IrSimpleFunction) {
+                val function = generateDelegatedFunction(subClass, irField, it, declarationStorage.findOverriddenFirFunction(it))
+                subClass.addMember(function)
+            } else if (it is IrProperty) {
+                generateDelegatedProperty(subClass, irField, it, declarationStorage.findOverriddenFirProperty(it))
+            }
+        }
+    }
+
+    private fun IrDeclaration.isDelegatable(): Boolean {
+        // TODO: replace checks on descriptors
```

Comment:
Please do not add descriptor usages. Today evening I will push the opt-in for descriptors in IR, and after it you will have a compiler error here. I think you can do all of it using IR elements themselves